### PR TITLE
build: adds flag for debugging the binary with delve

### PR DESCRIPTION
**Pull request by lkwg82**
…ry with delve

### Purpose

enables debugging with delve, see documentation [PR/366 in docs](https://github.com/syncthing/docs/pull/366)

`go run build.go -debugBinary build` (omits -ldflags)

### CODE REVIEW

File: build.go [hunk 1]
```diff
@@ -458,6 +461,16 @@ func build(target target, tags []string) {
 	if race {
 		args = append(args, "-race")
 	}
+
+	if debugBinary {
+		args = append(args, "-gcflags=-N -l")
+		// omit -ldflags because enabled will get
+		// `Could not launch program: decoding dwarf section info at offset 0x0: too short` on 'dlv exec ...'
+		// see https://github.com/derekparker/delve/issues/79
```
Comment (AudriusButkevicius) [inline build.go]: perhaps the comment should go above.

File: build.go [hunk 2]
```diff
@@ -358,6 +359,7 @@ func parseFlags() {
 	flag.StringVar(&extraTags, "tags", extraTags, "Extra tags, space separated")
 	flag.StringVar(&installSuffix, "installsuffix", installSuffix, "Install suffix, optional")
 	flag.StringVar(&pkgdir, "pkgdir", "", "Set -pkgdir parameter for `go build`")
+	flag.BoolVar(&debugBinary, "debugBinary", debugBinary, "enable not-optimzed binary to use with delve, set -gcflags='-N -l' and omit -ldflags")
```
Comment (AudriusButkevicius) [inline build.go]: Also, all tags are either no case `installsuffix` or dash separated, we don't have camel case anywhere.

**Comment by AudriusButkevicius** [comment]
@st-review lgtm

**Comment by st-review** [comment]
@AudriusButkevicius: Noted! Need another LGTM or explicit merge command.

**Comment by AudriusButkevicius** [comment]
Needs title changed as it gets merged.

**Comment by lkwg82** [comment]
retest?

**Comment by AudriusButkevicius** [comment]
@st-review rebuild

**Comment by calmh** [comment]
Yeah negatory on the camelcase, and @lkwg82 you should base pull requests on a branch in your fork, not a branch in the main repo. In this case it even causes a problem because the build script picks up the branch name and it's too long for the version string :)

**Comment by lkwg82** [comment]
@calmh ok

**Comment by lkwg82** [comment]
Should I move the PR to another repo+branch?

File: build.go [hunk 3]
```diff
@@ -358,6 +359,7 @@ func parseFlags() {
 	flag.StringVar(&extraTags, "tags", extraTags, "Extra tags, space separated")
 	flag.StringVar(&installSuffix, "installsuffix", installSuffix, "Install suffix, optional")
 	flag.StringVar(&pkgdir, "pkgdir", "", "Set -pkgdir parameter for `go build`")
+	flag.BoolVar(&debugBinary, "debug-binary", debugBinary, "enable not-optimzed binary to use with delve, set -gcflags='-N -l' and omit -ldflags")
```
Comment (calmh) [inline build.go]: “Create unoptimized...” in the description I think

File: build.go [hunk 4]
```diff
@@ -458,6 +461,16 @@ func build(target target, tags []string) {
 	if race {
 		args = append(args, "-race")
 	}
+
+	// omit -ldflags because enabled will get
+	// `Could not launch program: decoding dwarf section info at offset 0x0: too short` on 'dlv exec ...'
+	// see https://github.com/derekparker/delve/issues/79
+	if debugBinary {
+		args = append(args, "-gcflags=-N -l")
```
Comment (calmh) [inline build.go]: Make that `“-gcflags”, “-N -l”` so it’s handled the same as other flags

**Comment by calmh** [comment]
Nah leave this one as is but for next

**Comment by lkwg82** [comment]
> Same change in install() as well?

yes

File: build.go [hunk 5]
```diff
@@ -466,16 +463,12 @@ func build(target target, tags []string) {
 	// `Could not launch program: decoding dwarf section info at offset 0x0: too short` on 'dlv exec ...'
 	// see https://github.com/derekparker/delve/issues/79
 	if debugBinary {
-		args = append(args, "-gcflags=-N -l")
+		args = append(args, "-gcflags", gcflags())
```
Comment (calmh) [inline build.go]: Now you're complificating again, just making this the string literal is fine :)

**Comment by lkwg82** [comment]
I finished.

**Comment by calmh** [comment]
@st-review merge

build: Add support for debug-only binary (i.e., dlv)

**Comment by st-review** [comment]
:ok_hand: Merged as e6551c84859ea01f2bf4499ca0ba0275d76462c4. Thanks, @lkwg82!

**Pull request closed without merging**